### DISCUSSION TITLE
What does SampleFrames  means?

### DISCUSSION THREAD

**Discussion by nicozhou7** [open-mmlab/mmaction2#655] [Q&A] [answered]
What does SampleFrames actually mean in the code? I see the parameters are clip_len, frame_interval, and num_clips. Can anybody explain it to me? I am a newbie here :)

**Comment by innerlee** [comment]
@dreamerlin

**Comment by innerlee** [comment]
Does the doc offer any help? https://mmaction2.readthedocs.io/en/latest/api.html#mmaction.datasets.pipelines.SampleFrames

**Comment by irvingzhang0512** [comment, marked as answer]
`SampleFrames` defines sample strategy for input frames. Sample strategy is defined as `clip_len x frame_interval x num_clips`. There are two sample strategy, uniform sampling and dense sampling.

Dense strategy samples continuous frames. To be more specific, samples one frame every `frame_interval` frames and finally get `clip_len` frames.
Uniform strategy(`clip_len == 1`) devides all frames into `num_clips` parts and samples one frame from each part.

If there are 96 frames for one sample, frame id in range [1, 96].
Dense strategy demo: `32x2x1` **may** select no.`1, 3, 5, ..., 63` frames.
Uniform strategy demo:`1x1x8` divides 100 frames into [1, 12], [13, 24], [25, 36], [37, 48], ..., [85, 96] and **may** select no.`1, 13, 25, ..., 85` frames.

More details could refer to source code

**Reply by innerlee** [reply]
> More details could refer to source code

This is a documentation improvement request 😃

**Reply by irvingzhang0512** [reply]
> > More details could refer to source code
> 
> This is a documentation improvement request 😃

Actullay I tried to explain this function to my intern but failed... A lot of details and edge cases are involved. Told him to find out himself... :)

**Reply by nicozhou7** [reply]
So, if each of my input videos consists of 64 frames, what SampleFrames should I use?

**Reply by irvingzhang0512** [reply]
> So, if each of my input videos consists of 64 frames, what SampleFrames should I use?

It depends on a lot of things, i.e. which model you are using now, what action you are detecting, how much computer cost do you have. Overvall, recommended sample strategies are listed in config filename, i.e. `32x2x1` for I3D models and `1x1x8` for TSM

**Reply by irvingzhang0512** [reply]
BTW, 3D models don't support uniform sampling

**Reply by nicozhou7** [reply]
Actually, I am using R2+1D for detecting whether a person successfully catches a tennis ball or not. So, do you mean if I follow the recommended strategies for sample frames is better?

**Reply by irvingzhang0512** [reply]
> Actually, I am using R2+1D for detecting whether a person successfully catches a tennis ball or not. So, do you mean if I follow the recommended strategies for sample frames is better?

It's hard to say which model and which strategy is the best. You'd better try everything and find yourself. Recommended strategies have kinetics-400 pretrained mdoels, which will help you get better results. And tsm is my favourite

**Reply by nicozhou7** [reply]
Thank you for the reply, will try it then, can i know the reason why do you prefer tsm?

Regarding kinetics pretrained model, is there any tutorial in this API?

**Reply by irvingzhang0512** [reply]
TSM is faster than 3d models and accuracy is good enough for me.

You could refer to [model zoo](https://mmaction2.readthedocs.io/en/latest/recognition_models.html) and [Tutorial 2: Finetuning Models](https://mmaction2.readthedocs.io/en/latest/tutorials/2_finetune.html)

**Reply by nicozhou7** [reply]
Great, thank you !

**Comment by stephen-nju** [comment]
it may like sliding window on video frame sequence,the windows size is clip_len*fram_interval,the start index of the windows can computed by  num_clips.

**Comment by JasseurHadded1** [comment]
> Dense strategy demo: 32x2x1 may select no.1, 3, 5, ..., 63 frames

Does that mean if I have a video with 200 frames for example and I use `clip_len=32, frame_interval=2, num_clips=1` then we discard all frames from 65 to 200?
